### PR TITLE
OneCycle LR & RigL Experiments

### PR DESCRIPTION
**Pull request by mvacaporale**
### **RigL**
The PR includes utilities for global pruning by RigL:
* Methods to prune by weight and add by grad and associated tests 
* A `CosineDecayPruneScheduler` class for decaying the pruning rate throughout training
* HF Trainer `RigLCallback` to integrate these utilities and train SparseWeight models using RigL
* HF Trainer `PlotDensitiesCallback` to log the densities of each layer

Main Results:
| model                        | train loss  | eval loss      | perplexity |
|------------------------------|:-----------:|:--------------:|:----------:|
| tiny_bert_static_sparse_300k | 4.537       | 3.997          | 54.432     |
| tiny_bert_rigl_sparse_300k   | 5.963       | 5.774          | 321.95     |

^Note: These are sparse models.

### **OneCycleLR**
This PR also includes mixins for using the OneCycleLR scheduler and a LR Range Test mixin to help find good bounds for the max and min learning rates. The results are below. The max_lr was tuned tuned manually to be 0.01 while all the other params go by the default for the scheduler. 

| model                      | train loss  | eval loss      | perplexity |
|----------------------------|:-----------:|:--------------:|:----------:|
| tiny_bert_50k              | 5.990       | 5.800          | 330.234    |
| tiny_bert_one_cycle_lr_50k | 4.083       | 3.605          | 36.767     |

^Note: These are dense models.

TODO:
* Ideally, the `trainer_extra_kwargs` will be logged to wandb so that the lr params therein get logged. I'm currently looking into this and will add it in a future PR.

### CODE REVIEW

File: projects/transformers/callbacks/sparsity.py [hunk 1]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+        # Prune weights.
+        model.apply(rezero_weights)
+        prune_fraction = self.prune_scheduler.get_prune_fraction()
+        num_removed = global_prune_by_weight(self.sparse_modules, prune_fraction)
+        model.apply(rezero_weights)
+
+        # Post-prune sparsities.
+        param_sparsity1, mask_sparsity1 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Train for one step.
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:159]: This comment is a bit misleading and got me confused - this is not really training, just doing a forward/backward pass to calculate the grads. Training would imply there is a gradient step, but there isn't.

File: projects/transformers/callbacks/sparsity.py [hunk 2]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+        optimizer.zero_grad()
+        train_batch = next(iter(train_dataloader))
+        inputs_to_device(train_batch, device=args.device)
+        output = model(**train_batch)
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:163]: If compute_loss is not being used, this can't be combined with our distillation mixin.

File: projects/transformers/callbacks/sparsity.py [hunk 3]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+        output.loss.backward()
+
+        # Regrow weights
+        num_add = self.prune_scheduler.get_num_add(num_removed)
+        global_add_by_grad(self.sparse_modules, num_add)
+        self.prune_scheduler.step()
+
+        # Post-grow sparsities.
+        param_sparsity2, mask_sparsity2 = calc_cumulative_sparsity(self.sparse_modules)
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:172]: These params and mask vectors are only required for logging, as so other variables and computations in this method. I see the usefulness, but we will not need those once we tested the method and incorporate into our main experiments. To optimize for speed we could make all these optional in a future PR.

File: projects/transformers/callbacks/sparsity.py [hunk 4]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+            "rigl/actual_pruned_total_params": actual_pruned,
+            "rigl/pre_prune_param_sparsity": param_sparsity0,
+            "rigl/pre_prune_mask_sparsity": mask_sparsity0,
+            "rigl/post_prune_param_sparsity": param_sparsity1,
+            "rigl/post_prune_mask_sparsity": mask_sparsity1,
+            "rigl/pre_grow_param_sparsity": param_sparsity2,
+            "rigl/post_grow_mask_sparsity": mask_sparsity2,
+        })
+
+        wandb.log(logs, step=state.global_step)
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:196]: Will this fail if I am not using wandb, or this will just be ignored?

File: projects/transformers/callbacks/sparsity.py [hunk 5]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+        )
+        plot = wandb.Image(ax)
+        wandb.log({"delta_on_params": plot}, step=state.global_step)
+
+
+# -------------
+# Utilities
+# -------------
+
+def inputs_to_device(inputs, device):
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:286]: Just a heads up that if we switch to use `Trainer.prediction_step` there is already a call to `Trainer._prepare_inputs` which is the same as this.

File: projects/transformers/callbacks/sparsity.py [hunk 6]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+    Prep inputs dict by transferring all inputs to the device.
+    """
+    for k, v in inputs.items():
+        if isinstance(v, torch.Tensor):
+            inputs[k] = v.to(device)
+
+
+def calc_sparsity(tensor):
+    """Calculate the sparsity of a given tensor."""
+    num_zero = (tensor == 0).sum().item()
```
Comment (lucasosouza) [inline projects/transformers/callbacks/sparsity.py:297]: For all cases where we have `t == 0` or `t != 0`, a maybe faster alternative is `torch.count_nonzero`

File: projects/transformers/trainer_mixins/lr_range_test.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from torch.optim.lr_scheduler import LambdaLR
+
+
+class LRRangeTestMixin:
```
Comment (lucasosouza) [inline projects/transformers/trainer_mixins/lr_range_test.py:25]: I read the code and the docstring, but I gotta say have no idea how is this mixin calculating a LRRange or how it is supposed to be used.

File: projects/transformers/trainer_mixins/one_cycle_lr.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        """
+
+        # Set lr scheduler to dummy variable so it's not created in the call to super.
+        self.lr_scheduler = ...
+
+        # Create just the optimizer.
+        super().create_optimizer_and_scheduler(num_training_steps)
+
+        # Now define the lr scheduler, given the optimizer.
+        self.lr_scheduler = OneCycleLR(
```
Comment (lucasosouza) [inline projects/transformers/trainer_mixins/one_cycle_lr.py:73]: Instead of setting lr_scheduler, calling create_optimizer_and_scheduler and then redefining lr_scheduler, why not override `Trainer.create_scheduler` instead?

File: projects/transformers/trainer_mixins/one_cycle_lr.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    Mixin to HF Trainer for using the `OneCycleLR`_. See documentation for argument
+    details.
+
+    .. _OneCycleLR:
+        https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR
+    """
+
+    def __init__(
+        self,
+        max_lr=1e-2,
```
Comment (lucasosouza) [inline projects/transformers/trainer_mixins/one_cycle_lr.py:36]: We have to be careful that all these extra attributes are under trainer_kwargs, and they might start colliding. I prefixed the ones in KD with kd_. Not saying it is an ideal or prefered solution, just something to keep in mind.

**Review by lucasosouza — APPROVED** [review]
Left lots of comments, but one major issue I want to draw your attention to : this won't work with the distillation mixin. I am approving anyway so we can merge and move on, but this should be fixed before we can combine the mixins.

File: projects/transformers/callbacks/sparsity.py (hunk 1 above)
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:159]: Goatch. I'll change it to "Accumulate gradients over one batch." How's that?

File: projects/transformers/callbacks/sparsity.py (hunk 3 above)
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:172]: Good point. I'll add a todo.

File: projects/transformers/callbacks/sparsity.py (hunk 4 above)
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:196]: Good catch. It will fail. I'll fix it.

File: projects/transformers/callbacks/sparsity.py (hunk 5 above)
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:286]: Thanks for the heads up. It's an idempotent function. So it is fine to call more than once.

File: projects/transformers/trainer_mixins/one_cycle_lr.py (hunk 8 above)
Comment (mvacaporale) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: I don't believe there is such a class method, at least in version 4.4.2. There is `get_scheduler`, but that belongs to `transformers.optimization`.

File: projects/transformers/trainer_mixins/one_cycle_lr.py (hunk 9 above)
Comment (mvacaporale) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:36]: Fair point. We generally didn't do that in Vernon and we haven't run into issues yet. I'll keep the argument names as they are. But I'll keep it in mind moving forward.

File: projects/transformers/callbacks/sparsity.py [hunk 10]
```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
...
+            model, include_modules=[SparseWeightsBase]
+        ).values()
+
+    def on_step_end(
+        self, args, state, control, model=None,
+        train_dataloader=None, optimizer=None, **kwargs
+    ):
+        """Prune and regrow weights every 'prune_freq' iterations."""
+
+        if state.global_step % self.prune_freq != 0:
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:143]: It is in fact 0. I doubled checked. This kind of thing can be a tricky bug, so thanks for asking.

File: nupic/research/frameworks/dynamic_sparse/global_pruning.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+from nupic.research.frameworks.pytorch.mask_utils import get_topk_submask
+
+__all__ = [
+    "global_prune_by_weight",
+    "global_add_by_grad",
+]
+
+
+def global_prune_by_weight(sparse_modules, prune_fraction):
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dynamic_sparse/global_pruning.py:32]: I'll change it to `global_prune_by_abs_weight`, since that's technically the criteria.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dynamic_sparse/global_pruning.py:32]: Similarly, I've changed `global_add_by_grad` to `global_add_by_abs_grad`.

File: projects/transformers/callbacks/sparsity.py (hunk 6 above)
Comment (mvacaporale) [reply projects/transformers/callbacks/sparsity.py:297]: I've made the change in case it helps.

File: projects/transformers/trainer_mixins/lr_range_test.py (hunk 7 above)
Comment (mvacaporale) [reply projects/transformers/trainer_mixins/lr_range_test.py:25]: I've updated the docstring. Please let me know what you think. It's much more explicit in how to use it.

File: nupic/research/frameworks/dynamic_sparse/global_pruning.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    selecting the top absolute weights. Modules are pruned adjusting their `zero_masks`;
+    switching off the weight occurs by changing 0 to 1. Be sure to call
+    `rezero_weights` following this function to zero out the pruned weights.
+
+    :param sparse_modules: list of modules of type SparseWeightsBase
+    :param prune_fraction: fraction of weights to prune; between 0 and 1
+    """
+    assert 0 <= prune_fraction <= 1
+
+    # Flatten parameters to compare them all at once for global pruningn.
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dynamic_sparse/global_pruning.py]: > This won't work with the distillation mixin. I am approving anyway so we can merge and move on, but this should be fixed before we can combine the mixins.

Thanks for catching this. I'll make RigL into a mixin, seems it's the only way. But I'll leave it for a soon-to-come PR.

**Comment by mvacaporale** [comment]
> This won't work with the distillation mixin. I am approving anyway so we can merge and move on, but this should be fixed before we can combine the mixins.

Thanks for catching this. I'll make RigL into a mixin, seems it's the only way. But I'll leave it for a soon-to-come PR.

File: projects/transformers/callbacks/sparsity.py (hunk 1 above)
Comment (lucasosouza) [reply projects/transformers/callbacks/sparsity.py]: Sounds good

File: projects/transformers/trainer_mixins/one_cycle_lr.py (hunk 8 above)
Comment (lucasosouza) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: Interesting. There is such a function in 4.5
image

File: projects/transformers/trainer_mixins/lr_range_test.py (hunk 7 above)
Comment (lucasosouza) [reply projects/transformers/trainer_mixins/lr_range_test.py:25]: I am still confused on how to use it, maybe an example could help in a future PR. The example information might me more suited in an experiment file as well. We can talk about it later.

**Review by lucasosouza — APPROVED** [review]

File: projects/transformers/trainer_mixins/one_cycle_lr.py (hunk 8 above)
Comment (mvacaporale) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: I'm now on the cautious side for upgrading after seeing how quickly their repo changes. Last I tried 4.5.0, there was the cloudpickle issue when attempting a hyperparameter search. I might try it out in the next week or so, and we can discuss upgrading.

Comment (mvacaporale) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: Related: We should have tests for our mixins. As we upgrade the transformers version, things are likely to break if they keep refactoring the code as much as they do. I'll try and add some for onecycle lr, the lr range tests, and rigl that minimally invoke the trainer somehow. Hopefully, this could provide a template for your kd tests.

Comment (lucasosouza) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: Don't 100% agree with you. Yes, things are going to break, but we can also just fix them as they break. Tests are wonderful for products that require reliability, but they also take time, and in the past we've lost a significant amount of time building elaborate test suites for things we end up not using a few months later.

Comment (lucasosouza) [reply projects/transformers/trainer_mixins/one_cycle_lr.py:73]: I have some tests in my task list, for example, but the priority assigned to them is such that there is a small chance they make to the top of that list soon, unless our team grows fast or our deadlines are postponed.
